Applications/Spire/Source/UiViewer/DurationInputTestWidget.cpp

```diff
@@ -43,7 +45,7 @@ DurationInputTestWidget::DurationInputTestWidget(QWidget* parent)
   auto set_duration_button = make_flat_button(tr("Set Duration"), this);
   set_duration_button->setFixedHeight(scale_height(26));
   set_duration_button->connect_clicked_signal([=] { on_set_button(); });
-  layout->addWidget(set_duration_button, 2, 0, 1, 2);
+  layout->addWidget(set_duration_button, 3, 0, 1, 2);
```

Comment:
When I see the exact same code repeated 20 times it's a major indication that this design is wrong. What happens if another change is requested in the future and there are 40 widgets that need to be tested? Will that change need to be implemented 40 different times?

Propose a solution to this problem.

Reply:
Are you talking specifically about the initialization of the layouts?

Reply:
I'm talking about the overall pull request showing me 21 files with the same changes made to all of them. That is poor software development, it means that the cost of future changes to this project grow as a function of the total time spent on it. The purpose of using abstractions is to bring the cost of future changes from being linear down to being constant.

Reply:
I don't think it's possible to create a perfect abstraction, but after thinking it over I can create the grid layouts as sub-layouts of the main layout. At least that way if there's 'global' layout changes in the future they won't require cascading changes to the entire layout.

Reply:
The closest thing I can think of to this issue is something like an HTML template renderer. The problem is that aside from the 'Parameters' label the test widgets don't all share the same test parameters so there's no generic layout. The widgets that share functionality like the typed spin boxes and combo boxes already share common test widget layouts.

I can implement my suggestion of updating the layouts to be more flexible, but I'm not sure that another solution would be worth the time in coming up with and implementing.